File: python/feedo/modules/crypto.py
import os
import hmac
import hashlib
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from ecies import encrypt, decrypt
import binascii
from eth_account import Account

class FeedoCrypto:
    @staticmethod
    def generate_symmetric_key() -> bytes:
        return os.urandom(32)

    @staticmethod
    def derive_usage_key(wallet_private_key_hex: str) -> dict:
        """
        Deterministically derive the usage key (0xD) from the wallet key (0xW).

        usage_sk = HMAC-SHA256(key=wallet_sk, msg="feedo/usage-key/v1") mod n

        Returns {"private_key": "0x...", "address": "0x..."}.
        The derived key can sign requests but cannot move funds (USDT stay on 0xW).
        """
        sk_bytes = bytes.fromhex(wallet_private_key_hex.replace("0x", ""))
        digest = hmac.new(sk_bytes, b"feedo/usage-key/v1", hashlib.sha256).digest()
        usage_int = int.from_bytes(digest, "big")
        # secp256k1 group order n (reduction is negligible in practice, keeps the key valid)
        n = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141
        usage_int = usage_int % n
        if usage_int == 0:
            usage_int = 1
        usage_hex = format(usage_int, "064x")
        account = Account.from_key(usage_hex)
        return {"private_key": "0x" + usage_hex, "address": account.address}

    @staticmethod
    def encrypt_data(key: bytes, data: bytes) -> bytes:
        nonce = os.urandom(12)
        cipher = Cipher(algorithms.AES(key), modes.GCM(nonce))
        encryptor = cipher.encryptor()
        ciphertext = encryptor.update(data) + encryptor.finalize()
        return nonce + ciphertext + encryptor.tag

    @staticmethod
    def decrypt_data(key: bytes, encrypted_data: bytes) -> bytes:
        if len(encrypted_data) < 28:
            raise ValueError("Data is too short to be AES-GCM encrypted")
        nonce = encrypted_data[:12]
        tag = encrypted_data[-16:]
        ciphertext = encrypted_data[12:-16]
        
        cipher = Cipher(algorithms.AES(key), modes.GCM(nonce, tag))
        decryptor = cipher.decryptor()
        return decryptor.update(ciphertext) + decryptor.finalize()

    @staticmethod
    def encrypt_symmetric_key_ecies(public_key_hex: str, key: bytes) -> str:
        """
        public_key_hex: hex string of secp256k1 public key
        returns: hex string of the ECIES encrypted symmetric key
        """
        # Ensure it's uncompressed format if needed, eciespy accepts standard formats
        pub_bytes = binascii.unhexlify(public_key_hex.replace("0x", ""))
        encrypted = encrypt(public_key_hex, key)
        return binascii.hexlify(encrypted).decode('utf-8')

    @staticmethod
    def decrypt_symmetric_key_ecies(private_key_hex: str, encrypted_key_hex: str) -> bytes:
        """
        private_key_hex: hex string of secp256k1 private key
        encrypted_key_hex: hex string of the ECIES encrypted symmetric key
        returns: decrypted symmetric key bytes
        """
        enc_bytes = binascii.unhexlify(encrypted_key_hex.replace("0x", ""))
        return decrypt(private_key_hex, enc_bytes)

File: typescript/src/modules/crypto.ts
import { encrypt, decrypt } from 'eciesjs';
import * as crypto from 'crypto';
import { ethers } from 'ethers';

export class FeedoCrypto {
    static generateSymmetricKey(): Buffer {
        return crypto.randomBytes(32);
    }

    /**
     * Deterministically derive the usage key (0xD) from the wallet key (0xW).
     * usage_sk = HMAC-SHA256(key=wallet_sk, msg="feedo/usage-key/v1") mod n
     * The derived key can sign requests but cannot move funds (USDT stay on 0xW).
     */
    static deriveUsageKey(walletPrivateKeyHex: string): { privateKey: string; address: string } {
        const skBytes = Buffer.from(walletPrivateKeyHex.replace('0x', ''), 'hex');
        const digest = crypto.createHmac('sha256', skBytes).update('feedo/usage-key/v1').digest();
        let usageInt = BigInt('0x' + digest.toString('hex'));
        const n = BigInt('0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141');
        usageInt = usageInt % n;
        if (usageInt === 0n) usageInt = 1n;
        const usageHex = usageInt.toString(16).padStart(64, '0');
        const wallet = new ethers.Wallet('0x' + usageHex);
        return { privateKey: '0x' + usageHex, address: wallet.address };
    }

    static encryptData(key: Buffer, data: Buffer): Buffer {
        const nonce = crypto.randomBytes(12);
        const cipher = crypto.createCipheriv('aes-256-gcm', key, nonce);
        const ciphertext = Buffer.concat([cipher.update(data), cipher.final()]);
        const tag = cipher.getAuthTag();
        return Buffer.concat([nonce, ciphertext, tag]);
    }

    static decryptData(key: Buffer, encryptedData: Buffer): Buffer {
        if (encryptedData.length < 28) {
            throw new Error("Data is too short to be AES-GCM encrypted");
        }
        const nonce = encryptedData.subarray(0, 12);
        const tag = encryptedData.subarray(encryptedData.length - 16);
        const ciphertext = encryptedData.subarray(12, encryptedData.length - 16);

        const decipher = crypto.createDecipheriv('aes-256-gcm', key, nonce);
        decipher.setAuthTag(tag);
        return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
    }

    /**
     * @param publicKeyHex hex string of secp256k1 public key
     * @param key symmetric key bytes
     * @returns hex string of the ECIES encrypted symmetric key
     */
    static encryptSymmetricKeyEcies(publicKeyHex: string, key: Buffer): string {
        const pubKey = publicKeyHex.replace('0x', '');
        const encrypted = encrypt(pubKey, key);
        return Buffer.from(encrypted as any).toString('hex');
    }

    /**
     * @param privateKeyHex hex string of secp256k1 private key
     * @param encryptedKeyHex hex string of the ECIES encrypted symmetric key
     * @returns decrypted symmetric key bytes
     */
    static decryptSymmetricKeyEcies(privateKeyHex: string, encryptedKeyHex: string): Buffer {
        const privKey = privateKeyHex.replace('0x', '');
        const encBytes = Buffer.from(encryptedKeyHex.replace('0x', ''), 'hex');
        const decrypted = decrypt(privKey, encBytes);
        return Buffer.from(decrypted);
    }
}
